interfaces/builtin/wayland.go

```diff
@@ -157,9 +158,14 @@ func (iface *waylandInterface) UDevPermanentSlot(spec *udev.Specification, slot
 	return nil
 }
 
-func (iface *waylandInterface) AutoConnect(*snap.PlugInfo, *snap.SlotInfo) bool {
-	// allow what declarations allowed
-	return true
+func (iface *waylandInterface) AutoConnect(plug *snap.PlugInfo, slot *snap.SlotInfo) bool {
+  if release.OnClassic {
+	  // allow connection to implicit slot
+    return slot.Snap.GetType() == snap.TypeOS
+  } else {
+    // allow what declarations allowed
+    return true
+  }
```

Comment:
This is interesting because the base declaration is trying to handle this a bit:
```
  wayland:
    allow-installation:
      slot-snap-type:
        - app
        - core
    deny-connection:
      on-classic: false
    deny-auto-connection:
      on-classic: false
```

Which says to deny-auto-connection when not classic. Therefore, on classic distro, auto-connect. This is assuming that wayland on classic is coming from the host and not a snap. This probably should be adjusted to be:
```
  wayland:
    allow-installation:
      slot-snap-type:
        - app
        - core
    allow-connection:
      slot-snap-type:
      - core
    allow-auto-connection:
      slot-snap-type:
      - core
```
There are different ways of expressing this, but I think that this is the simplest. It also is a new variation of our documented patterns (see interfaces/policy/basedeclaration.go).

Reply:
> ```
>   wayland:
>     allow-installation:
>       slot-snap-type:
>         - app
>         - core
>     deny-connection:
>       on-classic: false
>     deny-auto-connection:
>       on-classic: false
> ```
> 
> 
> Which says to deny-auto-connection when not classic. Therefore, on classic distro, auto-connect.

I find these incantations hard to follow, but I think you've read this wrong. (Either that, or the current behaviour is broken.)

At present, on both Core and Classic, if there's a single wayland slot then plugs auto-connect. That is, auto-connect is **not** denied on "not classic", it happens. (E.g. https://tutorials.ubuntu.com/tutorial/secure-ubuntu-kiosk#2)

> This probably should be adjusted to be:
> 
> ```
>   wayland:
>     allow-installation:
>       slot-snap-type:
>         - app
>         - core
>     allow-connection:
>       slot-snap-type:
>       - core
>     allow-auto-connection:
>       slot-snap-type:
>       - core
> ```
> 
> There are different ways of expressing this, but I think that this is the simplest. It also is a new variation of our documented patterns (see interfaces/policy/basedeclaration.go).

Maybe I'm misreading your proposal. Wouldn't this prevent auto-connection to mir-kiosk on Core?

I want a change so that if there's an implicit slot (as is the case on Classic) that this takes precedence.

Reply:
I misread the initial proposal. I see now in your response and in https://forum.snapcraft.io/t/the-wayland-interface-and-auto-connect/13063/1 what you are trying to address is when there are multiple providing slots, that the classic one be chosen rather than have snapd refuse to connect any. Sorry about that.

(My comments were only considering changing the behavior of a single providing slot)

Removing my comment since it is confusing.